File: laplace/linresid2D_mpi_sync.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 6395
! 30 March 2009
!=================================================================


subroutine linresid2D(u, f, res, norm2, locN, locM, dx, dy, comm, ierr)
  !-----------------------------------------------------------------
  ! Description: 
  !    calculates the 2D linear residual 
  !             res = L*u - f
  !    and its L2-norm.
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi

  !======= Declarations =========
  implicit none

  integer, intent(in) :: locN, locM, comm
  integer, intent(out) :: ierr
  double precision, intent(in) :: dx, dy
  double precision, intent(in),  dimension(locN,locM) :: u, f
  double precision, intent(out), dimension(locN,locM) :: res
  double precision, intent(out) :: norm2
  double precision, dimension(locN) :: Nrecv, Srecv, Nsend, Ssend
  double precision, dimension(locM) :: Erecv, Wrecv, Esend, Wsend
  double precision :: tmp
  integer :: my_id, i, j, pdims(2), pcoords(2)
  logical :: periods(2)
  integer :: nbE, nbW, nbN, nbS, nbcoords(2)
  integer :: status(MPI_STATUS_SIZE)

  !======= Internals ============
  
  ! Get MPI parallelism information from comm
  call MPI_Cart_get(comm, 2, pdims, periods, pcoords, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) ' error in MPI_Cart_get = ', ierr
     call MPI_Abort(comm, 1, ierr)
  endif
  call MPI_Comm_rank(comm, my_id, ierr)
  if (ierr /= MPI_SUCCESS) then
     write(0,*) ' error in MPI_Comm_rank = ', ierr
     call MPI_Abort(comm, 1, ierr)
  endif

  ! initialize send/recv buffers
  Esend = u(locN,1:locM)
  Wsend = u(1,1:locM)
  Nsend = u(1:locN,locM)
  Ssend = u(1:locN,1)
  Nrecv = 0.d0
  Srecv = 0.d0
  Erecv = 0.d0
  Wrecv = 0.d0

  ! determine process 'neighbors'
  if (pcoords(1) /= 0) then
     nbcoords(1) = pcoords(1)-1
     nbcoords(2) = pcoords(2)
     call MPI_Cart_rank(comm, nbcoords, nbW, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) ' error in MPI_Cart_rank = ', ierr
        call MPI_Abort(comm, 1, ierr)
     endif
  end if
  if (pcoords(1) /= pdims(1)-1) then
     nbcoords(1) = pcoords(1)+1
     nbcoords(2) = pcoords(2)
     call MPI_Cart_rank(comm, nbcoords, nbE, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) ' error in MPI_Cart_rank = ', ierr
        call MPI_Abort(comm, 1, ierr)
     endif
  end if
  if (pcoords(2) /= 0) then
     nbcoords(1) = pcoords(1)
     nbcoords(2) = pcoords(2)-1
     call MPI_Cart_rank(comm, nbcoords, nbS, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) ' error in MPI_Cart_rank = ', ierr
        call MPI_Abort(comm, 1, ierr)
     endif
  end if
  if (pcoords(2) /= pdims(2)-1) then
     nbcoords(1) = pcoords(1)
     nbcoords(2) = pcoords(2)+1
     call MPI_Cart_rank(comm, nbcoords, nbN, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) ' error in MPI_Cart_rank = ', ierr
        call MPI_Abort(comm, 1, ierr)
     endif
  end if
  
  ! phase 1: even pcoords(1) send East; odd recv West
  if (modulo(pcoords(1),2) == 0) then
     if (pcoords(1) /= pdims(1)-1) then   ! check we're not last
        call MPI_Send(Esend, locM, MPI_REAL8, nbE, 100, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(1) /= 0) then            ! check we're not first
        call MPI_Recv(Wrecv, locM, MPI_REAL8, nbW, 100, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 2: even pcoords(1) recv East; odd send West
  if (modulo(pcoords(1),2) == 0) then
     if (pcoords(1) /= pdims(1)-1) then   ! check we're not last
        call MPI_Recv(Erecv, locM, MPI_REAL8, nbE, 101, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(1) /= 0) then            ! check we're not first
        call MPI_Send(Wsend, locM, MPI_REAL8, nbW, 101, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 3: odd pcoords(1) send East; even recv West
  if (modulo(pcoords(1),2) == 1) then
     if (pcoords(1) /= pdims(1)-1) then   ! check we're not last
        call MPI_Send(Esend, locM, MPI_REAL8, nbE, 102, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(1) /= 0) then            ! check we're not first
        call MPI_Recv(Wrecv, locM, MPI_REAL8, nbW, 102, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 4: odd pcoords(1) recv East; even send West
  if (modulo(pcoords(1),2) == 1) then
     if (pcoords(1) /= pdims(1)-1) then   ! check we're not last
        call MPI_Recv(Erecv, locM, MPI_REAL8, nbE, 103, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(1) /= 0) then            ! check we're not first
        call MPI_Send(Wsend, locM, MPI_REAL8, nbW, 103, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 5: even pcoords(2) send North; odd recv South
  if (modulo(pcoords(2),2) == 0) then
     if (pcoords(2) /= pdims(2)-1) then   ! check we're not last
        call MPI_Send(Nsend, locN, MPI_REAL8, nbN, 104, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(2) /= 0) then            ! check we're not first
        call MPI_Recv(Srecv, locN, MPI_REAL8, nbS, 104, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 6: even pcoords(2) recv North; odd send South
  if (modulo(pcoords(2),2) == 0) then
     if (pcoords(2) /= pdims(2)-1) then   ! check we're not last
        call MPI_Recv(Nrecv, locN, MPI_REAL8, nbN, 105, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(2) /= 0) then            ! check we're not first
        call MPI_Send(Ssend, locN, MPI_REAL8, nbS, 105, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 7: odd pcoords(2) send North; even recv South
  if (modulo(pcoords(2),2) == 1) then
     if (pcoords(2) /= pdims(2)-1) then   ! check we're not last
        call MPI_Send(Nsend, locN, MPI_REAL8, nbN, 106, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(2) /= 0) then            ! check we're not first
        call MPI_Recv(Srecv, locN, MPI_REAL8, nbS, 106, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if

  ! phase 8: odd pcoords(2) recv North; even send South
  if (modulo(pcoords(2),2) == 1) then
     if (pcoords(2) /= pdims(2)-1) then   ! check we're not last
        call MPI_Recv(Nrecv, locN, MPI_REAL8, nbN, 107, comm, status, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Recv = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  else
     if (pcoords(2) /= 0) then            ! check we're not first
        call MPI_Send(Ssend, locN, MPI_REAL8, nbS, 107, comm, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) ' error in MPI_Send = ', ierr
           call MPI_Abort(comm, 1, ierr)
        endif
     end if
  end if


  ! compute linear residual in interior of subdomain
  norm2 = 0.d0
  do j=2,locM-1
     do i=2,locN-1
        res(i,j) = (u(i-1,j) - 2.d0*u(i,j) + u(i+1,j))/dx/dx &
             + (u(i,j-1) - 2.d0*u(i,j) + u(i,j+1))/dy/dy - f(i,j)
        norm2 = norm2 + dx*dy*res(i,j)**2
     end do
  end do

  ! compute linear residual at South edge of local subdomain
  do i=2,locN-1
     res(i,1) = (u(i-1,1) - 2.d0*u(i,1) + u(i+1,1))/dx/dx &
          + (Srecv(i) - 2.d0*u(i,1) + u(i,2))/dy/dy - f(i,1)
     norm2 = norm2 + dx*dy*res(i,1)**2
  end do


  ! compute linear residual at West edge of local subdomain
  do j=2,locM-1
     res(1,j) = (Wrecv(j) - 2.d0*u(1,j) + u(2,j))/dx/dx           &
          + (u(1,j-1) - 2.d0*u(1,j) + u(1,j+1))/dy/dy - f(1,j)
     norm2 = norm2 + dx*dy*res(1,j)**2
  end do


  ! compute linear residual at East edge of local subdomain
  do j=2,locM-1
     res(locN,j) = (u(locN-1,j) - 2.d0*u(locN,j) + Erecv(j))/dx/dx &
          + (u(locN,j-1) - 2.d0*u(locN,j) + u(locN,j+1))/dy/dy - f(locN,j)
     norm2 = norm2 + dx*dy*res(locN,j)**2
  end do


  ! compute linear residual at North edge of local subdomain
  do i=2,locN-1
     res(i,locM) = (u(i-1,locM) - 2.d0*u(i,locM) + u(i+1,locM))/dx/dx &
          + (u(i,locM-1) - 2.d0*u(i,locM) + Nrecv(i))/dy/dy - f(i,locM)
     norm2 = norm2 + dx*dy*res(i,locM)**2
  end do


  ! compute linear residual at corners of local subdomain
  res(1,1) = (Wrecv(1) - 2.d0*u(1,1) + u(2,1))/dx/dx &
           + (Srecv(1) - 2.d0*u(1,1) + u(1,2))/dy/dy - f(1,1)
  norm2 = norm2 + dx*dy*res(1,1)**2

  res(locN,1) = (u(locN-1,1) - 2.d0*u(locN,1) + Erecv(1))/dx/dx &
       + (Srecv(locN) - 2.d0*u(locN,1) + u(locN,2))/dy/dy - f(locN,1)
  norm2 = norm2 + dx*dy*res(locN,1)**2

  res(1,locM) = (Wrecv(locM) - 2.d0*u(1,locM) + u(2,locM))/dx/dx &
       + (u(1,locM-1) - 2.d0*u(1,locM) + Nrecv(1))/dy/dy - f(1,locM)
  norm2 = norm2 + dx*dy*res(1,locM)**2

  res(locN,locM) = (u(locN-1,locM) - 2.d0*u(locN,locM) + Erecv(locM))/dx/dx &
       + (u(locN,locM-1) - 2.d0*u(locN,locM) + Nrecv(locN))/dy/dy - f(locN,locM)
  norm2 = norm2 + dx*dy*res(locN,locM)**2


  ! combine local sums into global L2-norm
  call MPI_Allreduce(norm2, tmp, 1, MPI_REAL8, MPI_SUM, comm, ierr)
  if (ierr /= 0) then
     print *,'linresid error: MPI_Allreduce failed with error = ',ierr
     call MPI_Abort(comm, 1, ierr)
  end if
  norm2 = sqrt(tmp)

  return

end subroutine linresid2D
!=================================================================

File: laplace/linresid2D_mpi_sync.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */

// Inclusions
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include "mpi.h"
#include "laplace2d.h"


/* Description:
      calculates the 2D linear residual
               res = L*u - f
      and its L2-norm. */
int linresid2D(double *u, double *f, double *res, double &norm,
	       int locN, int locM, double dx, double dy, MPI_Comm comm) {

  // get MPI parallelism information from comm
  int ierr, pdims[2], pcoords[2], periods[2];
  ierr = MPI_Cart_get(comm, 2, pdims, periods, pcoords);
  if (ierr != MPI_SUCCESS) {
    fprintf(stderr," error in MPI_Cart_get = %i\n",ierr);
    MPI_Abort(comm, 1);
  }

  /* allocate send/recv buffers */
  double *Esend = new double[locM];
  double *Wsend = new double[locM];
  double *Nsend = new double[locN];
  double *Ssend = new double[locN];
  double *Erecv = new double[locM];
  double *Wrecv = new double[locM];
  double *Nrecv = new double[locN];
  double *Srecv = new double[locN];

  /* initialize send/recv buffers */
  int i, j;
  for (j=0; j<locM; j++)  Esend[j] = u[idx(locN-1,j,locN)];
  for (j=0; j<locM; j++)  Wsend[j] = u[idx(0,j,locN)];
  for (i=0; i<locN; i++)  Nsend[i] = u[idx(i,locM-1,locN)];
  for (i=0; i<locN; i++)  Ssend[i] = u[idx(i,0,locN)];
  for (j=0; j<locM; j++)  Erecv[j] = 0.0;
  for (j=0; j<locM; j++)  Wrecv[j] = 0.0;
  for (i=0; i<locN; i++)  Nrecv[i] = 0.0;
  for (i=0; i<locN; i++)  Srecv[i] = 0.0;

  // determine process 'neighbors'
  int nbW, nbE, nbS, nbN, nbcoords[2];
  if (pcoords[0] > 0) {
    nbcoords[0] = pcoords[0]-1;
    nbcoords[1] = pcoords[1];
    ierr = MPI_Cart_rank(comm, nbcoords, &nbW);
    if (ierr != MPI_SUCCESS) {
      fprintf(stderr," error in MPI_Cart_rank = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }
  if (pcoords[0] < pdims[0]-1) {
    nbcoords[0] = pcoords[0]+1;
    nbcoords[1] = pcoords[1];
    ierr = MPI_Cart_rank(comm, nbcoords, &nbE);
    if (ierr != MPI_SUCCESS) {
      fprintf(stderr," error in MPI_Cart_rank = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }
  if (pcoords[1] > 0) {
    nbcoords[0] = pcoords[0];
    nbcoords[1] = pcoords[1]-1;
    ierr = MPI_Cart_rank(comm, nbcoords, &nbS);
    if (ierr != MPI_SUCCESS) {
      fprintf(stderr," error in MPI_Cart_rank = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }
  if (pcoords[1] < pdims[1]-1) {
    nbcoords[0] = pcoords[0];
    nbcoords[1] = pcoords[1]+1;
    ierr = MPI_Cart_rank(comm, nbcoords, &nbN);
    if (ierr != MPI_SUCCESS) {
      fprintf(stderr," error in MPI_Cart_rank = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  // phase 1: even pcoords[0] send East, odd recv West
  MPI_Status status;
  if (pcoords[0]%2 == 0) {
    if (pcoords[0] < pdims[0]-1) {
      ierr = MPI_Send(Esend, locM, MPI_DOUBLE, nbE, 100, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[0] > 0) {
      ierr = MPI_Recv(Wrecv, locM, MPI_DOUBLE, nbW, 100, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 2: even pcoords[0] recv East, odd send West
  if (pcoords[0]%2 == 0) {
    if (pcoords[0] < pdims[0]-1) {
      ierr = MPI_Recv(Erecv, locM, MPI_DOUBLE, nbE, 101, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[0] > 0) {
      ierr = MPI_Send(Wsend, locM, MPI_DOUBLE, nbW, 101, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 3: odd pcoords[0] send East, even recv West
  if (pcoords[0]%2 == 1) {
    if (pcoords[0] < pdims[0]-1) {
      ierr = MPI_Send(Esend, locM, MPI_DOUBLE, nbE, 102, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[0] > 0) {
      ierr = MPI_Recv(Wrecv, locM, MPI_DOUBLE, nbW, 102, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 4: odd pcoords[0] recv East, even send West
  if (pcoords[0]%2 == 1) {
    if (pcoords[0] < pdims[0]-1) {
      ierr = MPI_Recv(Erecv, locM, MPI_DOUBLE, nbE, 103, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[0] > 0) {
      ierr = MPI_Send(Wsend, locM, MPI_DOUBLE, nbW, 103, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 5: even pcoords[1] send North, odd recv South
  if (pcoords[1]%2 == 0) {
    if (pcoords[1] < pdims[1]-1) {
      ierr = MPI_Send(Nsend, locN, MPI_DOUBLE, nbN, 104, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[1] > 0) {
      ierr = MPI_Recv(Srecv, locN, MPI_DOUBLE, nbS, 104, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 6: even pcoords[1] recv North, odd send South
  if (pcoords[1]%2 == 0) {
    if (pcoords[1] < pdims[1]-1) {
      ierr = MPI_Recv(Nrecv, locN, MPI_DOUBLE, nbN, 105, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[1] > 0) {
      ierr = MPI_Send(Ssend, locN, MPI_DOUBLE, nbS, 105, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 7: odd pcoords[1] send North, even recv South
  if (pcoords[1]%2 == 1) {
    if (pcoords[1] < pdims[1]-1) {
      ierr = MPI_Send(Nsend, locN, MPI_DOUBLE, nbN, 106, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[1] > 0) {
      ierr = MPI_Recv(Srecv, locN, MPI_DOUBLE, nbS, 106, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // phase 8: odd pcoords[1] recv North, even send South
  if (pcoords[1]%2 == 1) {
    if (pcoords[1] < pdims[1]-1) {
      ierr = MPI_Recv(Nrecv, locN, MPI_DOUBLE, nbN, 107, comm, &status);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Recv = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  } else {
    if (pcoords[1] > 0) {
      ierr = MPI_Send(Ssend, locN, MPI_DOUBLE, nbS, 107, comm);
      if (ierr != MPI_SUCCESS) {
	fprintf(stderr," error in MPI_Send = %i\n",ierr);
	MPI_Abort(comm, 1);
      }
    }
  }

  // compute linear residual in interior of subdomain
  double mynorm = 0.0;
  for (i=1; i<locN-1; i++) {
    for (j=1; j<locM-1; j++) {
      res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
	+ (u[idx(i-1,j,locN)] - 2.0*u[idx(i,j,locN)] + u[idx(i+1,j,locN)])/dx/dx
	+ (u[idx(i,j-1,locN)] - 2.0*u[idx(i,j,locN)] + u[idx(i,j+1,locN)])/dy/dy;
      mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];
    }
  }

  // compute linear residual at South edge of local subdomain
  j=0;
  for (i=1; i<locN-1; i++) {
    res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
      + (u[idx(i-1,j,locN)] - 2.0*u[idx(i,j,locN)] + u[idx(i+1,j,locN)])/dx/dx
      + (Srecv[i] - 2.0*u[idx(i,j,locN)] + u[idx(i,j+1,locN)])/dy/dy;
    mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];
  }

  // compute linear residual at West edge of local subdomain
  i=0;
  for (j=1; j<locM-1; j++) {
    res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
      + (Wrecv[j] - 2.0*u[idx(i,j,locN)] + u[idx(i+1,j,locN)])/dx/dx
      + (u[idx(i,j-1,locN)] - 2.0*u[idx(i,j,locN)] + u[idx(i,j+1,locN)])/dy/dy;
    mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];
  }

  // compute linear residual at East edge of local subdomain
  i=locN-1;
  for (j=1; j<locM-1; j++) {
    res[idx(i,j,locN)] = -f[idx(i,j,locN)]
      + (u[idx(i-1,j,locN)] - 2.0*u[idx(i,j,locN)] + Erecv[j])/dx/dx
      + (u[idx(i,j-1,locN)] - 2.0*u[idx(i,j,locN)] + u[idx(i,j+1,locN)])/dy/dy ;
    mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];
  }

  // compute linear residual at North edge of local subdomain
  j=locM-1;
  for (i=1; i<locN-1; i++) {
    res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
      + (u[idx(i-1,j,locN)] - 2.0*u[idx(i,j,locN)] + u[idx(i+1,j,locN)])/dx/dx
      + (u[idx(i,j-1,locN)] - 2.0*u[idx(i,j,locN)] + Nrecv[i])/dy/dy;
    mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];
  }

  // compute linear residual at corners of local subdomain
  i=0; j=0;
  res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
    + (Wrecv[j] - 2.0*u[idx(i,j,locN)] + u[idx(i+1,j,locN)])/dx/dx
    + (Srecv[i] - 2.0*u[idx(i,j,locN)] + u[idx(i,j+1,locN)])/dy/dy;
  mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];

  i=locN-1; j=0;
  res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
    + (u[idx(i-1,j,locN)] - 2.0*u[idx(i,j,locN)] + Erecv[j])/dx/dx
    + (Srecv[i] - 2.0*u[idx(i,j,locN)] + u[idx(i,j+1,locN)])/dy/dy;
  mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];

  i=0; j=locM-1;
  res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
    + (Wrecv[j] - 2.0*u[idx(i,j,locN)] + u[idx(i+1,j,locN)])/dx/dx
    + (u[idx(i,j-1,locN)] - 2.0*u[idx(i,j,locN)] + Nrecv[i])/dy/dy;
  mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];

  i=locN-1; j=locM-1;
  res[idx(i,j,locN)] =  -f[idx(i,j,locN)]
    + (u[idx(i-1,j,locN)] - 2.0*u[idx(i,j,locN)] + Erecv[j])/dx/dx
    + (u[idx(i,j-1,locN)] - 2.0*u[idx(i,j,locN)] + Nrecv[i])/dy/dy;
  mynorm += dx*dy*res[idx(i,j,locN)]*res[idx(i,j,locN)];


  // combine local sums into global L2-norm
  ierr = MPI_Allreduce(&mynorm, &norm, 1, MPI_DOUBLE, MPI_SUM, comm);
  if (ierr != MPI_SUCCESS) {
    printf(" error in MPI_Allreduce = %i\n",ierr);
    MPI_Abort(comm, 1);
  }
  norm = sqrt(norm);


  // delete temporary buffers
  delete[] Esend;
  delete[] Wsend;
  delete[] Nsend;
  delete[] Ssend;
  delete[] Erecv;
  delete[] Wrecv;
  delete[] Nrecv;
  delete[] Srecv;

  return ierr;

} // end linresid
